### PR TITLE
Hierarchy and labeler

### PR DESCRIPTION
**Pull request by jambran**
Comments about our algorithm for labeling clusters.  Dig deep on this one, I'm not super happy with my explanations. Could also switch over to the powerpoint and just run through that

### CODE REVIEW

File: clustering_tutorial.ipynb [hunk 1]
```diff
@@ -1009,6 +1011,90 @@
...
+    "If each cluster had a _label_, we could sort on that label and get things closer together, focusing on the clusters we want to refine first. \n",
+    "\n",
+    "In order to label these clusters, we need a resource to know which words tend to get this label. This is the _hierarchy_, a tree of terms. \n",
+    "\n",
+    "Each node cruciall has\n",
+    "- a name or label, like `root-feeling-negative_sentiment-angry`\n",
+    "\n",
+    "and each node _might_ have \n",
+    "- children\n",
+    "- parents\n",
```
Comment (kirsten-stallings) [inline clustering_tutorial.ipynb]: "A child of `root-feeling-negative_sentiment- angry` might be `root-feeling-negative_sentiment- angry-irate`"
"The parent of `root-feeling-negative_sentiment- angry` is `root-feeling-negative_sentiment`"

File: clustering_tutorial.ipynb [hunk 2]
```diff
@@ -1009,6 +1011,90 @@
...
+    "- words that are associated with this label, like `upset, frustrated, in_denial, angry`\n",
+    "\n",
+    "Each child of that node contains concepts that are more specific. \n",
+    "\n",
+    "For instance, \n",
+    "> dog -> herders -> shetland sheepdog\n",
+    "\n",
+    "Given a cluster and a word embedding model, the _labeler_ moves through the hierarchy to determine the best label for this unseen cluster.\n",
```
Comment (kirsten-stallings) [inline clustering_tutorial.ipynb]: a section on the different methods we used to determine how we chose the best label?

File: clustering_tutorial.ipynb [hunk 3]
```diff
@@ -1009,6 +1011,90 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## The Labeler\n",
+    "\n",
+    "The labeler moves throught the hierarchy to find the best fitting name for the unseen cluster generated above in the clustering step. \n",
+    "\n",
+    "First, it checks each child of root. For each node, we find the vectors for each word present in that node _and_ its children using - and this is important! - the individual client's FastText model. \n",
```
Comment (kirsten-stallings) [inline clustering_tutorial.ipynb]: I kind of like subtree better than children. Children makes me think exactly one deeper

We look at two centroids:
1. The vectors for the words in the node and it's subtree
2. The vectors for only the words in the node.

Comment (kirsten-stallings) [inline clustering_tutorial.ipynb]: Each "node in the subtree of root"?

File: clustering_tutorial.ipynb [hunk 4]
```diff
@@ -1009,6 +1011,90 @@
...
+    "\n",
+    "Quiz: _Why is it important to use the client's model here?_\n",
+    "\n",
+    "We average these vectors to get the _centroid_. The serves as the middle location for this node in the hierarchy. For each child of root, we have a centroid. \n",
+    "\n",
+    "Next, we calculate the centroid for the unseen cluster. Which of the child-of-root centroids falls closest to the centroid of the unseen cluster? \n",
+    "\n",
+    "We repeat this process until we find the closest node-centroid in the hierarchy.\n",
```
Comment (kirsten-stallings) [inline clustering_tutorial.ipynb]: tie back into cosim and euclidean distance?

File: clustering_tutorial.ipynb [hunk 5]
```diff
@@ -1009,6 +1011,90 @@
...
+    "![moving through the hierarchy](figures/moving-through-the-hierarchy.gif)\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# we can't show you any code for this - we'd have to build a whole new GOT hierarchy"
```
Comment (kirsten-stallings) [inline clustering_tutorial.ipynb]: can show visualizer since we'll be there but i'd leave this comment in for the future.

**Comment by kirsten-stallings** [comment]
I left comments on what i would change. I'm also not against using the powerpoint and going to the visualizer. We can add the powerpoint to the git repo and link it in the notebook

File: clustering_tutorial.ipynb (hunk 1 above)
Comment (jambran) [reply clustering_tutorial.ipynb]: Good call, thanks!

File: clustering_tutorial.ipynb (hunk 3 above)
Comment (jambran) [reply clustering_tutorial.ipynb]: This is great, thanks!

Comment (jambran) [reply clustering_tutorial.ipynb]: I ended up restructuring this with your other change - let me know what you think

File: clustering_tutorial.ipynb (hunk 4 above)
Comment (jambran) [reply clustering_tutorial.ipynb]: Oh, yeah, good idea!

**Pull request merged**